jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,128 @@
+        ('ngev', int),  # total number of jacobian evaluations
+        ('nhev', int),  # total number of hessian evaluations
+        ('x_k', jnp.ndarray),
+        # A tensor containing the last argument value found during the search. If the search converged, then
+        # this value is the argmin of the objective function.
+        ('f_k', jnp.ndarray),  # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        ('g_k', jnp.ndarray),  # A tensor containing the gradient of the
+        # objective function at the
+        # `final_position`. If the search converged
+        # the max-norm of this tensor should be
+        # below the tolerance.
+        ('H_k', jnp.ndarray)  # A tensor containing the inverse of the estimated Hessian.
+    ])
+
+
+def bfgs_minimize(func, x0, options=None):
+    """
+        Minimisation with BFGS. Inexact line search is performed satisfying both Wolfe conditions
+        (strong version of second one).
+        Args:
+            func: function to minimise, should take a single flat parameter, similar to numpy
+            x0: initial guess of parameter
+            options: A dictionary of solver options. All methods accept the following generic options:
+                maxiter : int or None which means inf
+                    Maximum number of iterations to perform. Depending on the
+                    method each iteration may use several function evaluations.
+                analytic_initial_hessian: bool,
+                    whether to use JAX to get hessian for the initialisation, otherwise it's eye.
+                g_tol: float
+                    stopping criteria is |grad(func)|_2 \lt g_tol
+                ls_maxiter: int or None which means inf
+                    Maximum number of line search iterations to perform
+
+        Returns: BFGSResults
+
+    """
+    if options is None:
+        options = dict()
+    maxiter: Optional[int] = options.get('maxiter', None)
+    analytic_initial_hessian: bool = options.get('analytic_initial_hessian', True)
+    g_tol: float = options.get('g_tol', 1e-8)
+    ls_maxiter: int = options.get('ls_maxiter', 50)
+
+    state = BFGSResults(converged=False,
+                        failed=False,
+                        k=0,
+                        nfev=0,
+                        ngev=0,
+                        nhev=0,
+                        x_k=x0,
+                        f_k=None,
+                        g_k=None,
+                        H_k=None)
+
+    if maxiter is None:
+        maxiter = jnp.inf
+
+    D = x0.shape[0]
+
+    if analytic_initial_hessian:
+        hess = jax.hessian(func, argnums=0)
+        initial_B = hess(x0)
+        initial_H = jnp.linalg.pinv(initial_B)
+        state = state._replace(nhev=state.nhev + 1)
```

Comment:
I'm not sure that it actually makes sense to compute analytic Hessians just because JAX makes it _possible_. If this isn't an established method (e.g., something found in Nocedal's book), then I would lean against supporting it.